### DISCUSSION TITLE
Training a Vocoder using TTS outputs

### DISCUSSION THREAD

**Discussion by saiakarsh193** [coqui-ai/TTS#3470] [General Q&A]
I am trying to train a `HiFiGAN` vocoder for a custom dataset from scratch. I already trained a `FastPitch` TTS model for this dataset. The `Synthesizer` module uses Griffin-Lim by default which does not give great audio quality. So here I can train a vocoder using the same dataset and the issue is fixed. 

However, it is a common practice to use the synthesized TTS outputs as the training data for the vocoder. This gives the vocoder a better idea of how the TTS is generating mel spec and hence gives better quality. But I realised that the mel spec generated by TTS is different from ground truth, not just in values but also in shape. There is no guarantee (especially in Auto-Regressive models), that the output length is the same as the input length. Therefore we have an issue where the generated TTS mel spec shape is incorrect for the given hop length, and window length. For example, 

```
sample 1: 
gt (80, 418)
syn (80, 342)

sample 2:
gt (80, 652)
syn (80, 87)

sample 3:
gt (80, 796)
syn (80, 610)
```
you can see that the generated mel spec (`syn`) has a different shape to that of the ground truth (`gt`) and in some cases (like in `sample 2`), the shapes are drastically different (as a result of TTS quality). 

If I ignore this and just go ahead and train, then the [code will throw an error](https://github.com/coqui-ai/TTS/blob/dev/TTS/vocoder/datasets/gan_dataset.py#L134) (as it should!)

So is there a way of fixing this issue, or do I have no option but to use the ground truth for training the vocoder?

**Comment by p0p4k** [comment]
Yes, train end to end, search for a paper called JETS.
